### ISSUE TITLE
Allow users to provide user/case-specific user plugins as input

### ISSUE THREAD

**Issue by ntouran** [terrapower/armi#569]
There are situations where users may want to provide case-specific material properties, component shapes, etc. that are typically provided by plugins. Thus, it'd be nice if users could point to a plugin that resides in an input file using a new setting (e.g. `userPluginPath`). 

Since it's related to the overall repeatability of the run, the new file should be hashed in the stdout and copied into the output h5 file for safe keeping. 

This will also close #535

Sidenote: our pre-existing custom Python input related to the `shuffleLogic` and `controlLogic`  settings could likely be merged into this capability at a later time to make this an overall cleanup.

I'll take a stab at implementing this.

**Comment by ntouran** [comment]
Hmm, well maybe. See [comment here](https://github.com/terrapower/armi/issues/535#issuecomment-1033305726)

**Comment by jakehader** [comment]
I like the thoughts on the other ticket for the materials, but do we still want this for the other plugins too (fuel management/control rod logic)?

**Comment by jakehader** [comment]
After talking it sounds like we just need a one line change in the Case to add the working directory of the case it `sys.path` and the custom materials will work. Maybe we can add an example case of this somewhere in the gallery too?

**Comment by jakehader** [comment]
@ntouran are you able to make a PR for this?

**Comment by ntouran** [comment]
Yeah so after more discussion it's clear that this kind of thing would still be useful for some purposes, including doing rapid prototyping of test code that does things at special hooks. Effectively, this would allow people to use the plugin system as a convenient pathway for monkey patching during development/testing. I think as long as we put a strong indication near the ARMI version information that user plugins are in use, then we can maintain knowledge in the stdout about what code was actually run. Obviously, app-endorsed plugins explicitly imported and registered by the application should generally be used for actual qualified/traceable runs and analysis.

**Comment by ntouran** [comment]
K so in the PR linked above I fixed this with environmental variables. Using user settings would be perhaps nicer from a UI perspective, but is challenging from an architectural point of view because plugins must be loaded and locked down before settings can be parsed (since settings are defined by plugins...)

**Comment by sombrereau** [comment]
@keckler -- FYI this might help you configure a reactor without shuffling logic or control logic

**Comment by keckler** [comment]
Yeah, I agree, thanks @sombrereau . _If_ it ever happens 😆

**Comment by john-science** [comment]
> Yeah, I agree, thanks @sombrereau . _If_ it ever happens laughing

I mean, it happened like a month ago.
